### ISSUE TITLE
autoUpdater.quitAndInstall fails to relaunch on Mac OS 11 Big Sur

### ISSUE THREAD

**Issue by kklash** [electron/electron#25626]
### Preflight Checklist
* [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x] I have searched the issue tracker for an issue that matches the one I want to file, without success.

### Issue Details

I wanted to post this in https://github.com/electron/electron/issues/24319 but the issue is locked.

[Related comment with more details](https://github.com/Squirrel/Squirrel.Mac/issues/251#issuecomment-698010499).

* **Electron Version:**
  * v8.5.1
* **Operating System:**
  * Mac OS 11 Big Sur Beta
* **Last Known Working Electron version:**
  * n/a

### Actual Behavior

When restarting an upgraded electron application with [autoUpdater.quitAndInstall](https://www.electronjs.org/docs/api/auto-updater#autoupdaterquitandinstall), the application quits and upgrades successfully, but is NOT relaunched thereafter.

### Expected Behavior

The application should be launched after upgrading.

### To Reproduce

It's very hard for me to reproduce this issue because Squirrel.Mac requires that any MacOS app using auto-update is code-signed, and I can't figure out how to build Squirrel from source successfully so as to artificially remove that restriction. 

If there are any guides available for building Squirrel.Mac + Electron together, i would be very thankful to hear about them.

### Additional Information

As mentioned [here](https://github.com/Squirrel/Squirrel.Mac/issues/251#issuecomment-698010499), there are two lines in the Squirrel install log normally present with earlier versions of MacOS which are missing on the Big Sur log.

```
2020-09-17 17:55:15.344 ShipIt[4099:100431] Application launched at file:///Applications/Whatever.app/
2020-09-17 17:55:15.344 ShipIt[4099:100431] ShipIt quitting
```

Those lines are logged from these locations:

https://github.com/Squirrel/Squirrel.Mac/blob/44468f858ce0d25c27bd5e674abfa104e0119738/Squirrel/ShipIt-main.m#L128
https://github.com/Squirrel/Squirrel.Mac/blob/44468f858ce0d25c27bd5e674abfa104e0119738/Squirrel/ShipIt-main.m#L145

My original plan was to add some extra logging statements in Squirrel.Mac's source code and use that to determine why those lines are missing, but i hit a brick wall trying to build Squirrel and Electron together from lack of documentation.

**Comment by MarshallOfSound** [comment]
Hey @kklash I've been trying to figure this one out for a while 😅  just never opened an issue so thanks for that, it's definitely a weird one and unfortunately I haven't really got anything to go on.

> If there are any guides available for building Squirrel.Mac + Electron together, i would be very thankful to hear about them.

The latest Electron `master` branch builds Squirrel.Mac inline from source, if you follow our build instructions the Squirrel.Mac source code will be at `{checkout}/src/third_party/squirrel.mac` and you can make whatever code changes you like there.

**Comment by kklash** [comment]
Thanks @MarshallOfSound, i'll take a look at that tomorrow and see what I can manage =)

**Comment by MarshallOfSound** [comment]
@kklash I'm pretty sure this is a bug in macOS itself, trying to find a workaround but might have to raise a TSI with apple 🤔

**Comment by nbashkankov** [comment]
Hello, I got same issue for Electron 7.2.1 for Big Sur
quitAndInstall - only updates App and not relaunched with new version:

@MarshallOfSound @kklash 
probably this details could help you to find where is the issue:

grabbed from "Console" for "ShipIt" during restart phase

>default	| 17:03:23.016700+0300 |	powerd	| Process ShipIt.2295 Released PreventSystemSleep "Updating" age:00:00:00  id:30064805569 [System: PrevSleep DeclUser kCPU kDisp]
default	17:03:23.017254+0300	ShipIt	Installation completed successfully
default	17:03:23.017569+0300	powerd	Process ShipIt.2295 Released PreventSystemSleep "Updating" age:00:00:00  id:30064805570 [System: DeclUser kDisp]
default	17:03:23.039699+0300	ShipIt	Initializing connection
default	17:03:23.039828+0300	ShipIt	Removing all cached process handles
default	17:03:23.039912+0300	ShipIt	**Sending handshake request attempt 1 to server**
default	17:03:23.039953+0300	ShipIt	**Creating connection to com.apple.runningboard**
error	17:03:23.040860+0300	ShipIt	**Connection interrupted: will attempt to reconnect**
default	17:03:23.040918+0300	ShipIt	**Handshake in progress; aborting new handshake attempt**
error	17:03:23.040982+0300	ShipIt	**XPC message send failed**
error	17:03:23.041131+0300	ShipIt	**Handshake failed with error: <Error Domain=NSCocoaErrorDomain Code=4101 "Couldn’t communicate with a helper application.">**
default	17:03:23.041210+0300	ShipIt	Sending handshake request attempt 2 to server
default	17:03:23.041230+0300	ShipIt	Creating connection to com.apple.runningboard
error	17:03:23.042601+0300	ShipIt	Connection interrupted: will attempt to reconnect
default	17:03:23.042670+0300	ShipIt	Handshake in progress; aborting new handshake attempt
error	17:03:23.042776+0300	ShipIt	XPC message send failed
error	17:03:23.042862+0300	ShipIt	Handshake failed with error: <Error Domain=NSCocoaErrorDomain Code=4101 "Couldn’t communicate with a helper application.">
default	17:03:23.042929+0300	ShipIt	Sending handshake request attempt 3 to server

and such attempts repeats many times.

**Comment by nbashkankov** [comment]
Any news here? maybe it possible to do some temporary workaround outside squirrel ?

**Comment by kklash** [comment]
We implemented a stopgap measure which adjusts the wording in our app. Normally, after downloading an app update, we show the user a button that says 'Restart Now' but instead we adjust that wording if the user is on big sur, saying 'update installed, please quit and restart the app to apply the changes'

**Comment by nbashkankov** [comment]
@kklash - with your solution you will end up with launched ShipIt process which will constantly repeat launch attempts - even if your app already restarted. you can checkout it by watch over "Console" filtered by "ShipIt" and see ShipIt process in Activity Monitor.

**Comment by nbashkankov** [comment]
Btw, looks like `setBadgeCount` - https://www.electronjs.org/docs/api/app#appsetbadgecountcount-linux-macos
 also not working on BigSur.

**Comment by twitharshil** [comment]
@MarshallOfSound Any chance that the fix will be backported to the supported versions - v9 or v10?

**Comment by MarshallOfSound** [comment]
@twitharshil Unfortunately not, it requires several build system changes that aren't safe enough to backport

**Comment by alectrocute** [comment]
@MarshallOfSound Thank you for fixing this in v11.

**Comment by koenoe** [comment]
Is this fix yet released? Having this problem with Electron 11.0.3

**Comment by abmai** [comment]
same ^, still having this issue in Electron 11.0.3

**Comment by alectrocute** [comment]
@koenoe @abmai Please confirm it's not a `BrowserWindow` preventing the quit and install with something like:

`js
function callBeforeQuitAndInstall() {
  try {
    const electron = require("electron");
    const app = electron.app;
    const BrowserWindow = electron.BrowserWindow;
    if (tray) tray = null;
    app.removeAllListeners("window-all-closed");
    var browserWindows = BrowserWindow.getAllWindows();
    browserWindows.forEach(function(browserWindow) {
      browserWindow.removeAllListeners("close");
    });
  } catch (e) {
    console.log(e);
  }
}`

**Comment by koenoe** [comment]
Still having this issue in 11.0.4. Also tried @alectrocute suggestion, no effect.
Works fine on Windows.

**Comment by koenoe** [comment]
Still having this issue in 11.1.0
Any idea when this will be looked at? Or do I need to create a new issue?

**Comment by MarshallOfSound** [comment]
To be clear for folks in this thread, the fix only works for updating from 11.x to 11.x, i.e. your app must already be on 11.x in order for the updater to correctly relaunch.  If your users are on older Electron versions they will have to manually restart the app (the update itself worked fine).  This can't be worked around as the updater itself needed to be updated, so folks running an older updater to update to the newer version **aren't running the fix yet**.

TLDR: Trying updating from 11.X to 11.X and you'll see it work perfectly fine

**Comment by koenoe** [comment]
Yes, that's what I'm doing. We have a test build with a fake app version (99.99.99) that has same Electron version as our RC.
It works fine on Windows and Mac OS Catalina, but not on Big Sur. If I'm the only one with the problem, I'll do some more digging on our side, but looks like there's more people.

**Comment by ivanggq** [comment]
My test updating an app with Electron 11 did not work. I can see the new code in ShipIt being executed, because I see that ShipIt prints this log:
```
Attempting to launch app on lower than 11.0
```
This is the changed code in ShipIt, which checks for macOS version 11 and performs conditional launch code:
```
if (@available(macOS 11.0, *)) {
	NSLog(@"Attempting to launch app on 11.0 or higher");
	NSTask *task = [[NSTask alloc] init];
	[task setLaunchPath: @"/usr/bin/open"];
	[task setArguments: [NSArray arrayWithObjects: bundleURL.path, nil]];
	[task launch];
	[task waitUntilExit];
	NSLog(@"Application launched at %@", bundleURL);
} else {
	NSLog(@"Attempting to launch app on lower than 11.0");
	if (![NSWorkspace.sharedWorkspace launchApplicationAtURL:bundleURL options:NSWorkspaceLaunchDefault configuration:@{} error:&error]) {
		NSLog(@"Could not launch application at %@: %@", bundleURL, error);
		return;
	}
	NSLog(@"Application launched at %@", bundleURL);
}
```
The `if (@available(macOS 11.0, *))` check is failing and executing the launch code that does not work on macOS 11.

**Comment by alectrocute** [comment]
> Yes, that's what I'm doing. We have a test build with a fake app version (99.99.99) that has same Electron version (11.x) as our RC.
> It works fine on Windows and Mac OS Catalina, but not on Big Sur. If I'm the only one with the problem, I'll do some more digging on our side, but looks like there's more people.

@koenoe Greetings! Did you and your team ever determine the cause of this issue? Was it Xcode SDK-related?

**Comment by mohammedshoib** [comment]
It worked for me after updating electron builder 11.1.1.
Note:  Current electron app version & the autoupdate version , both should have latest electron builder i.e. 11.1.1

**Comment by koenoe** [comment]
> > Yes, that's what I'm doing. We have a test build with a fake app version (99.99.99) that has same Electron version (11.x) as our RC.
> > It works fine on Windows and Mac OS Catalina, but not on Big Sur. If I'm the only one with the problem, I'll do some more digging on our side, but looks like there's more people.
> 
> @koenoe Greetings! Did you and your team ever determine the cause of this issue? Was it Xcode SDK-related?

Yes, we're using a Castlabs fork and they were building with old Xcode SDK. See: https://github.com/castlabs/electron-releases/issues/86

**Comment by twitharshil** [comment]
For the folks who don't want to upgrade their application to electron v11 just to fix this auto-updater issue, they can simply replace the squirrel.mac pre-build binaries downloaded by electron before compiling the electron artifacts with the binary build on top of latest commit for [Squirrel.mac](https://github.com/Squirrel/Squirrel.Mac/commits/master)[ at least work for electron v9 ](I would expect squirrel.mac to release the latest binary with all the fixes as their latest release is built on top of a very old commit)
P.S: You need to build Squirrel.mac binary with the latest Xcode versions (currently the latest version is 12.3) so as to make the workaround work on bigSur OS mentioned by @ivanggq in this [comment](https://github.com/electron/electron/issues/25626#issuecomment-756442879)

**Comment by alectrocute** [comment]
Excellent info @twitharshil, thank you

**Comment by Denz1994** [comment]
@ivanggq Regarding your change in https://github.com/electron/electron/issues/25626#issuecomment-756442879: 

How did you manage to get your change into your electron app? It looks like [electron/DEPS](https://github.com/electron/electron/blob/master/DEPS) defaults to the commit at `squirrel.mac_version`.

**Comment by ivanggq** [comment]
@Denz1994 The issue for us got fixed after we moved to electron 12 and also we built it with Xcode 12 - not too sure which one of these helped.

**Comment by Denz1994** [comment]
Understood, sadly that isn't an option in my case. 

@twitharshil Is there electron support for replacing the squirrel. pre-built binaries? I don't see anything in the electron/electron-builder documentation. Did you just make the change in shipIt-main.m, build squirrel.mac, then include it as an `extraResource` in your package.json?

Just looking for some direction, into pushing this squirrel.mac patch into my application.

**Comment by MarshallOfSound** [comment]
> Just looking for some direction, into pushing this squirrel.mac patch into my application, without having to upgrade.

For stability reasons, there's no _safe_ way to hotswap a shared library like Squirrel.Mac.  Sometimes you can just copy in a new build and the symbol tables happen to line up.  It's not going to be 100% safe though.  I'd strongly suggest just upgrading Electron versions.  Is there anything holding you back from doing that?

**Comment by Denz1994** [comment]
Following up on my previous comment. I can vouch that updating from electron 9.1.2 to electron 11.4.0 worked in my case. More specifically, the related upgrades were as follows:
```
 "electron": "11.4.0",
 "electron-builder": "22.9.1",
 "electron-updater": "4.3.5",
```

Thanks, @MarshallOfSound for your response.
